crates/sui-core/src/authority_active.rs

```diff
@@ -176,33 +181,185 @@ impl<A> ActiveAuthority<A>
+        let checkpoint_join = receiver.spawn("Gossip".to_string(), move || {
+            let inner_locals = checkpoint_locals.clone();
+            async move {
+                if checkpoint {
+                    checkpoint_process(&inner_locals, &CheckpointProcessControl::default()).await;
+                }
+                Ok(())
             }
         });
 
-        if let Err(err) = _gossip_join.await {
-            error!("Join gossip task end error: {:?}", err);
-        }
-        if let Err(err) = _checkpoint_join.await {
-            error!("Join checkpoint task end error: {:?}", err);
+        join!(gossip_join, checkpoint_join);
+    }
+}
+
+/*
+    The active authority start-update-shutdown facilities.
+
+    The active authority runs in a separate task, and manages the tasks of separate active
+    processes, with one sub-task per active process. These subtasks internally listen to a
+    watch channel on which others may send a start, shutdown, or update signal. When this
+    is received the tasks are restarted, or shut down.
+
+*/
+
+use tokio::sync::watch;
+
+pub enum LifecycleSignal {
+    Start,
+    Restart,
+    Exit,
+}
+
+pub struct LifecycleSignalSender {
+    sender: watch::Sender<LifecycleSignal>,
+}
+
+impl LifecycleSignalSender {
+    pub fn new() -> (Self, LifecycleTaskHandler) {
+        let (sender, receiver) = watch::channel(LifecycleSignal::Start);
+        (
+            LifecycleSignalSender { sender },
+            LifecycleTaskHandler { receiver },
+        )
+    }
+
+    pub async fn signal(&self, signal: LifecycleSignal) {
+        // TODO: reflect on what to do with this error.
+        let _ = self.sender.send(signal);
+    }
+}
+
+#[derive(Clone)]
+pub struct LifecycleTaskHandler {
+    pub receiver: watch::Receiver<LifecycleSignal>,
+}
+
+impl LifecycleTaskHandler {
+    pub async fn spawn<F, T>(mut self, description: String, fun: F)
+    where
+        F: Fn() -> T,
+        T: Future + Send + 'static,
+        <T as Future>::Output: Borrow<Result<(), SuiError>> + Send + 'static,
+    {
+        let mut handle: Option<tokio::task::JoinHandle<Result<(), SuiError>>> = None;
+        loop {
+            // flags for each iteration on whether we should
+            // start, restart or exit the task.
+            let mut start = false;
+            let mut abort = false;
+            let mut exit = false;
+
+            let signal = self.receiver.changed().await;
+            match signal {
+                Err(_err) => {
+                    info!("Closing active tasks, command channel was dropped.");
+                    abort = true;
+                    exit = true;
+                }
+                Ok(()) => {
+                    // We have actually received a new signal.
+                    match *self.receiver.borrow() {
+                        LifecycleSignal::Start => {
+                            info!("Starting task: {}", description);
```

Comment:
I think most of these tracing lines should probably be lower level than info to avoid polluting the logs, maybe debug or trace?

Reply:
now `debug`